utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -457,130 +523,163 @@ pub enum State {
+					transport: uri.to_owned().into(),
+					state_snapshot: None,
+					pallets: pallet.clone(),
+					child_trie: *child_tree,
+					hashed_keys: vec![
+						// we always download the code, but we almost always won't use it, based on
+						// `Runtime`.
+						well_known_keys::CODE.to_vec(),
+						// we will always download this key, since it helps detect if we should do
+						// runtime migration or not.
+						[twox_128(b"System"), twox_128(b"LastRuntimeUpgrade")].concat(),
+						[twox_128(b"System"), twox_128(b"Number")].concat(),
+					],
+					hashed_prefixes: vec![],
+				}))
 			},
-		})
-	}
+		};
+
+		// possibly overwrite the state version, should hardly be needed.
+		let builder = if let Some(state_version) = shared.overwrite_state_version {
+			log::warn!(
+				target: LOG_TARGET,
+				"overwriting state version to {:?}, you better know what you are doing.",
+				state_version
+			);
+			builder.overwrite_state_version(state_version)
+		} else {
+			builder
+		};
+
+		// then, we prepare to replace the code based on what the CLI wishes.
+		let maybe_code_to_overwrite = match shared.runtime {
+			Runtime::Path(ref path) => Some(std::fs::read(path).map_err(|e| {
+				format!("error while reading runtime file from {:?}: {:?}", path, e)
+			})?),
+			Runtime::Existing => None,
+		};
+
+		// build the main ext.
+		let mut ext = builder.build().await?;
+
+		// actually replace the code if needed.
+		if let Some(new_code) = maybe_code_to_overwrite {
+			let original_code = ext
+				.execute_with(|| sp_io::storage::get(well_known_keys::CODE))
+				.expect("':CODE:' is always downloaded in try-runtime-cli; qed");
+
+			// NOTE: see the impl notes of `read_runtime_version`, the ext is almost not used here,
+			// only as a backup.
+			ext.insert(well_known_keys::CODE.to_vec(), new_code.clone());
+			use parity_scale_codec::Decode;
+			let old_version = <RuntimeVersion as Decode>::decode(
+				&mut &*executor.read_runtime_version(&original_code, &mut ext.ext()).unwrap(),
+			)
+			.unwrap();
+			log::info!(
+				target: LOG_TARGET,
+				"original spec: {:?}-{:?}",
+				old_version.spec_name,
+				old_version.spec_version
+			);
+			let new_version = <RuntimeVersion as Decode>::decode(
+				&mut &*executor.read_runtime_version(&new_code, &mut ext.ext()).unwrap(),
+			)
+			.unwrap();
+			log::info!(
+				target: LOG_TARGET,
+				"new spec: {:?}-{:?}",
+				new_version.spec_name,
+				new_version.spec_version
+			);
+
+			if new_version.spec_name != old_version.spec_name {
+				return Err("Spec names must match.".into())
+			}
+		}
 
-	/// Get the uri, if self is `Live`.
-	pub(crate) fn live_uri(&self) -> Option<String> {
-		match self {
-			State::Live { uri, .. } => Some(uri.clone()),
-			_ => None,
+		// whatever runtime we have in store now must have been compiled with try-runtime feature.
+		if !ensure_try_runtime::<Block, HostFns>(&shared, &ext) {
+			return Err("given runtime is NOT compiled with try-runtime feature!".into())
 		}
```

Comment:
```suggestion
		if runtime_version.api_version(TryRuntime::ID).is_none() {
			return Err("given runtime is NOT compiled with try-runtime feature!".into())
		}
```

Something like that is much better and easier :P

Reply:
Beautiful sir.

Reply:
Well, we don't work at the `client/ProvideRuntimeApi` level here, so hard to find something that implements `ApiExt` from the state machine or the raw code. Might give up.

Reply:
yeah, it seems like all such functionality is only available if you build a whole client, which we totally don't want to do here. https://paritytech.github.io/substrate/master/sp_api/trait.ProvideRuntimeApi.html#implementors

Reply:
That is the reason I had written `runtime_version` :P